### DISCUSSION TITLE
Earning bug fixed?

### DISCUSSION THREAD

**Discussion by Sparwen** [JK-Rolling/DuinoI2C_ESP#1] [Q&A] [answered]
I´m wondering if the "bug" with low earning over I2C vs USB is fixed in this version?
Please see old reported bug at https://github.com/ricaun/DuinoCoinI2C/issues/90.

Will I earn as much DUCOs with say 10 arduino I2C slaves as with 10 arduinos mining with AVR-miner?

Regards

**Comment by JK-Rolling** [comment, marked as answer]
I wouldn't say it's a "bug", more like a software architecture limitation.

In this new DuinoI2C_ESP, I re-architected the software from scratch to
fully resolve this.

In my beta tester program, I did receive feedback that the earning is
similar to USB.

You may unlock the firmware by referring to Unlock Key
<https://github.com/JK-Rolling/DuinoI2C_ESP/tree/main/esp8266#unlock-key> which
allows up to 20 workers!

**Reply by Sparwen** [reply]
Okey so it's simular to USB but not the same?

I guess I need to test it out for myself then.

I hope it's light years away from the performance like the old code?😁
It was terrible, wasn't even an improvement when adding slaves? Just an increased in power consumption...

**Reply by JK-Rolling** [reply]
I'm looking forward for your test result. Do share them when you have it :)

**Reply by Sparwen** [reply]
Sorry for the late answer :)

I did some short term testing ( approximately 5500shares, about 1,5-2h of mining )
And for 7pcs of i2c slaves vs 7pcs of USB miners, there were only about **-11%** decrease in performance ( 55.33 vs 62.49 DUCS/day )

This is a major improvement compered to the old code!  👍 

I will share later on for long-term and bigger scale testing.

Keep up the good work!

**Reply by JK-Rolling** [reply]
thanks for the encouraging data!
recent experiment done (by jpx13) shows further 8% improvement if the I2C master is using 100KHz while the worker (ATtiny85) is using 400KHz.
![image](https://github.com/user-attachments/assets/48fe697e-979d-48bc-bf8f-ea270af1ee71)


One may also need to consider the power consumption between both setup VS the Duco earning difference too.
USB - Host + USB Hub + workers
I2C - Master + workers

**Reply by Sparwen** [reply]
Interesting!

I thought that the i2c slaves had to be on the same frequency as the i2c master? 
But apparently not :D 

There seems to be more or less no differens in performance regarding Shares ( **-0.5%** )

- 7pcs USB miners: 460680 shares for 24h 
- 7pcs I2C miners: 458439 shares for 24h

But there still seems to be a slightly decrease in DUCO erning performance ( **about -6% to -7%** )

- 7pcs USB miners: 64.29 DUCOs/day
- 7pcs I2C miners: 59.55 DUCOs/day

That´s still awsome compared to the old code 👍

**Reply by JK-Rolling** [reply]
awesome comparison! 
The earning performance is outside of the ESP control, but maybe it is justified in terms of power saving for ESP :-)

btw, I assume you're not running both USB and I2C simultaneously as Kolka dictates efficiency drops for worker that join later

**Reply by Sparwen** [reply]
No, exactly, I did not run them simultaneously. They run separately 24 hours apart :)

**Reply by JK-Rolling** [reply]
do you mind sharing more info on the setup? the table, data 7 and 8 is missing a few info
https://github.com/JK-Rolling/DuinoI2C_ESP/discussions/5#discussioncomment-11069281

**Comment by CGameDev** [comment]
Regarding the d1 mini, in the firmware can you have the option of having the onboard led be passed to one of the digital PINs as output and allow it to be turned on in the web interface. Similar to the Pico code where i change the code that i can us a LED on the PCB rather than the onBoard Pico led.  I would like to design a PCB that have the ESP light on the board.
![image](https://github.com/user-attachments/assets/5612db57-64d5-4e17-88aa-f80c37bb9f78)
6 or 7, i think those pins dont start high or low when starting up the d1 mini, i think they are refer to as "safe" pins
![image](https://github.com/user-attachments/assets/0d4a7907-1aa4-4b15-b082-f32d57632e6a)

**Reply by JK-Rolling** [reply]
without changing anything, pin D4 is already sharing pin with the builtin LED.
for dedicated GPIO13/D7, please use this customized [DuinoI2C_ESP.ino.d1_mini.400K.ledD7.v0.20.1.bin.gz](https://github.com/user-attachments/files/17090080/DuinoI2C_ESP.ino.d1_mini.400K.ledD7.v0.20.1.bin.gz) to work with Pico worker
